File: packages/core/src/normalized-cache.ts
import type { TypePolicy, NormalizedCacheConfig } from './types';

// #######################################
// CONSTANTS
// #######################################

/** Default upper bound on the number of stored entities before LRU eviction. */
const DEFAULT_MAX_ENTITIES = 5000;

// #######################################
// TYPE DEFINITIONS
// #######################################

/**
 * A reference to a normalized entity in the cache.
 * Format: "TypeName:id"
 */
export interface EntityRef {
  __ref: string;
}

// #######################################
// TYPE GUARDS
// #######################################

function isEntityRef(value: unknown): value is EntityRef {
  return (
    typeof value === 'object' &&
    value !== null &&
    '__ref' in value &&
    typeof (value as EntityRef).__ref === 'string'
  );
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

// #######################################
// NORMALIZED CACHE
// #######################################

/**
 * Normalized entity store for GraphQL responses.
 *
 * Entities are identified by their __typename + key field(s) and stored in a
 * flat map ("User:42" → fields). This simplified entity store focuses solely on
 * entity management, with query coordination handled by QueryClient.
 *
 * Objects that lack __typename (or whose type has no resolvable key) pass
 * through unchanged, so non-GraphQL queries are unaffected.
 */
export class NormalizedCache {
  // ##############################
  // STATE
  // ##############################

  /** Flat entity store: "Type:id" → merged fields */
  private entities = new Map<string, Record<string, unknown>>();

  /** Reverse index: entity ref → query keys that reference it */
  private entityToQueries = new Map<string, Set<string>>();

  /** Callbacks invoked when a specific entity changes (for useFragment) */
  private entityListeners = new Map<string, Set<() => void>>();

  private config: NormalizedCacheConfig;

  /** Soft upper bound on stored entities; LRU eviction keeps us at or below it. */
  private maxEntities: number;

  /**
   * Wired up by QueryClient: reports whether a query key currently has active
   * subscribers. Used to "pin" entities that a live view depends on so LRU
   * eviction never pulls data out from under a mounted component.
   */
  isQueryKeyActive?: (queryKey: string) => boolean;

  // ##############################
  // INITIALIZATION
  // ##############################

  constructor(config: NormalizedCacheConfig = {}) {
    this.config = config;
    // A non-positive or NaN cap is meaningless (it would evict everything or
    // nothing coherently), so fall back to the default. `Infinity` is a valid
    // opt-out and survives this check.
    this.maxEntities =
      typeof config.maxEntities === 'number' && config.maxEntities > 0
        ? config.maxEntities
        : DEFAULT_MAX_ENTITIES;
  }

  // ##############################
  // PUBLIC API: SUBSCRIPTIONS
  // ##############################

  /**
   * Subscribe to changes for a specific entity.
   * The callback is invoked whenever `writeFragment` or `evict` touches this entity.
   * Returns an unsubscribe function.
   */
  subscribeToEntity(
    typename: string,
    id: string | number,
    callback: () => void,
  ): () => void {
    const ref = this.makeRef(typename, id);
    this.ensureSet(this.entityListeners as Map<string, Set<unknown>>, ref).add(
      callback,
    );
    return () => {
      this.entityListeners.get(ref)?.delete(callback);
    };
  }

  // ##############################
  // PUBLIC API: NORMALIZATION
  // ##############################

  /**
   * Normalize a value and store entities in the entity store.
   * Returns the normalized shape with entity refs.
   * The queryKey is used to build the reverse index (entity → queries).
   */
  normalize(data: unknown, queryKey: string): unknown {
    const shape = this.normalizeValue(data, queryKey);
    this.enforceCapacity();
    return shape;
  }

  /** Current number of entities held in the store. */
  size(): number {
    return this.entities.size;
  }

  /**
   * Reconstruct denormalized data from a normalized shape.
   */
  denormalize(shape: unknown): unknown {
    return this.denormalizeValue(shape);
  }

  // ##############################
  // PUBLIC API: FRAGMENT OPERATIONS
  // ##############################

  /**
   * Update a specific entity. Merges `data` into the existing record (shallow
   * merge of top-level fields) and returns the query keys that reference it.
   * The caller (QueryClient) is responsible for notifying affected queries.
   */
  writeFragment(
    typename: string,
    id: string | number,
    data: Record<string, unknown>,
  ): string[] {
    const ref = this.makeRef(typename, id);
    const existing = this.entities.get(ref) ?? {};
    const merged = this.mergeEntity(typename, existing, data);
    this.setEntity(ref, merged);
    this.notifyEntityListeners(ref);
    const affected = this.getAffectedQueries(ref);
    this.enforceCapacity();
    return affected;
  }

  /**
   * Read a raw (non-denormalized) entity record from the store.
   * Returns undefined if the entity is not cached.
   */
  readFragment<T extends Record<string, unknown> = Record<string, unknown>>(
    typename: string,
    id: string | number,
  ): T | undefined {
    return this.entities.get(this.makeRef(typename, id)) as T | undefined;
  }

  /**
   * Remove an entity from the store and return the query keys that were
   * referencing it (so the caller can invalidate them).
   */
  evict(typename: string, id: string | number): string[] {
    const ref = this.makeRef(typename, id);
    const affected = this.getAffectedQueries(ref);
    this.entities.delete(ref);
    this.entityToQueries.delete(ref);
    this.notifyEntityListeners(ref);
    return affected;
  }

  // ##############################
  // INTERNAL HELPERS: ENTITY REFS
  // ##############################

  private makeRef(typename: string, id: string | number): string {
    return `${typename}:${id}`;
  }

  /**
   * Derive the entity ref for an object using its __typename and key field(s).
   * Returns null if the object cannot be identified (no __typename, no key).
   */
  private getEntityRef(obj: Record<string, unknown>): string | null {
    const typename = obj.__typename;
    if (typeof typename !== 'string') return null;

    const policy: TypePolicy | undefined = this.config.typePolicies?.[typename];
    const keyFields = policy?.keyFields ?? 'id';

    let keyValue: unknown;
    if (Array.isArray(keyFields)) {
      const parts = keyFields.map((f) => obj[f]);
      if (parts.some((p) => p == null)) return null;
      keyValue = parts.join(':');
    } else {
      keyValue = obj[keyFields as string];
    }

    if (keyValue == null) return null;
    return this.makeRef(typename, keyValue as string | number);
  }

  // ##############################
  // INTERNAL HELPERS: ENTITY MERGING
  // ##############################

  /**
   * Merge new data into an existing entity using the type's merge policy.
   * Falls back to shallow merge if no custom merge function is defined.
   */
  private mergeEntity(
    typename: string,
    existing: Record<string, unknown>,
    incoming: Record<string, unknown>,
  ): Record<string, unknown> {
    const policy = this.config.typePolicies?.[typename];
    return policy?.merge
      ? policy.merge(existing, incoming)
      : { ...existing, ...incoming };
  }

  // ##############################
  // INTERNAL HELPERS: INDEX MANAGEMENT
  // ##############################

  /**
   * Ensure a Set exists at the given key in a Map, creating it if needed.
   * Returns the Set for chaining.
   */
  private ensureSet<K>(map: Map<K, Set<unknown>>, key: K): Set<unknown> {
    if (!map.has(key)) {
      map.set(key, new Set());
    }
    return map.get(key)!;
  }

  /**
   * Get the list of query keys that reference a specific entity.
   * Returns an empty array if the entity has no references.
   */
  private getAffectedQueries(ref: string): string[] {
    return Array.from(this.entityToQueries.get(ref) ?? []);
  }

  // ##############################
  // INTERNAL HELPERS: LRU EVICTION
  // ##############################

  /**
   * Insert or update an entity, marking it most-recently-used.
   *
   * A JS `Map` preserves insertion order and `set` on an existing key leaves
   * its position untouched, so we delete-then-set to move the entry to the
   * "newest" end. That gives us an O(1) LRU ordering straight from the Map:
   * the hashmap + doubly-linked-list you'd hand-roll for an LRU cache, except
   * the linked list is the engine's own.
   */
  private setEntity(ref: string, value: Record<string, unknown>): void {
    this.entities.delete(ref);
    this.entities.set(ref, value);
  }

  /** Mark an already-stored entity as most-recently-used (no-op if absent). */
  private touch(ref: string): void {
    const value = this.entities.get(ref);
    if (value !== undefined) {
      this.entities.delete(ref);
      this.entities.set(ref, value);
    }
  }

  /**
   * An entity is "pinned" — never eligible for eviction — when a live view
   * depends on it: either a query with active subscribers references it, or a
   * `useFragment`-style entity listener is subscribed directly to it. Evicting a
   * pinned entity would blank out mounted UI, so we keep it even over the cap.
   *
   * We deliberately do NOT invalidate/refetch on eviction: every query retains
   * its own complete `state.data` snapshot, so an inactive query whose entity is
   * evicted still renders correct cached data and simply repopulates the shared
   * store on its next fetch. Refetching here would be wasted work and, when the
   * cap is saturated by pinned entities, could loop.
   */
  private isPinned(ref: string): boolean {
    if ((this.entityListeners.get(ref)?.size ?? 0) > 0) return true;
    if (!this.isQueryKeyActive) return false;
    const queries = this.entityToQueries.get(ref);
    if (!queries) return false;
    for (const key of queries) {
      if (this.isQueryKeyActive(key)) return true;
    }
    return false;
  }

  /**
   * Evict least-recently-used, unpinned entities until the store is back within
   * `maxEntities`. If every over-cap entity is pinned the store stays above the
   * cap (soft limit) rather than break a live view.
   */
  private enforceCapacity(): void {
    if (this.entities.size <= this.maxEntities) return;

    // `keys()` yields least-recently-used first. Deleting the current/visited
    // key mid-iteration is well-defined for Map, so evicting in place is safe.
    for (const ref of this.entities.keys()) {
      if (this.entities.size <= this.maxEntities) break;
      if (this.isPinned(ref)) continue;
      this.entities.delete(ref);
      this.entityToQueries.delete(ref);
    }
  }

  // ##############################
  // INTERNAL HELPERS: OBJECT TRANSFORMATION
  // ##############################

  /**
   * Transform all values in an object using the given transformation function.
   * Returns a new object with the same keys but transformed values.
   */
  private transformObjectValues(
    obj: Record<string, unknown>,
    transform: (value: unknown) => unknown,
  ): Record<string, unknown> {
    const result: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(obj)) {
      result[k] = transform(v);
    }
    return result;
  }

  // ##############################
  // CORE LOGIC: NORMALIZATION
  // ##############################

  /** Recursively normalize a value, registering entity refs under queryKey. */
  private normalizeValue(value: unknown, queryKey: string): unknown {
    // ####################
    // Handle arrays
    // ####################

    if (Array.isArray(value)) {
      return value.map((item) => this.normalizeValue(item, queryKey));
    }

    // ####################
    // Handle primitives and non-plain objects
    // ####################

    if (!isPlainObject(value)) return value;

    // ####################
    // Normalize nested fields
    // ####################

    const normalized = this.transformObjectValues(value, (v) =>
      this.normalizeValue(v, queryKey),
    );

    // ####################
    // Check if this object is an identifiable entity
    // ####################

    const ref = this.getEntityRef(value);
    if (ref === null) {
      // Not an identifiable entity — store inline
      return normalized;
    }

    // ####################
    // Store entity and build reverse index
    // ####################

    const existing = this.entities.get(ref) ?? {};
    const typename = value.__typename as string;
    const merged = this.mergeEntity(typename, existing, normalized);
    this.setEntity(ref, merged);

    // Track which query references this entity
    this.ensureSet(this.entityToQueries as Map<string, Set<unknown>>, ref).add(
      queryKey,
    );

    return { __ref: ref } satisfies EntityRef;
  }

  // ##############################
  // CORE LOGIC: DENORMALIZATION
  // ##############################

  /** Recursively reconstruct a value by resolving entity refs. */
  private denormalizeValue(value: unknown): unknown {
    // ####################
    // Handle arrays
    // ####################

    if (Array.isArray(value)) {
      return value.map((item) => this.denormalizeValue(item));
    }

    // ####################
    // Handle primitives and non-plain objects
    // ####################

    if (!isPlainObject(value)) return value;

    // ####################
    // Resolve entity references
    // ####################

    if (isEntityRef(value)) {
      const entity = this.entities.get(value.__ref);
      if (entity === undefined) return undefined; // entity was evicted
      // Reading an entity counts as "using" it, so refresh its LRU recency.
      this.touch(value.__ref);
      return this.denormalizeValue(entity);
    }

    // ####################
    // Recursively denormalize nested objects
    // ####################

    return this.transformObjectValues(value, (v) => this.denormalizeValue(v));
  }

  // ##############################
  // INTERNAL HELPERS: NOTIFICATIONS
  // ##############################

  /** Invoke direct entity listeners (e.g. useFragment subscribers). */
  private notifyEntityListeners(ref: string): void {
    const listeners = this.entityListeners.get(ref);
    if (!listeners) return;
    for (const cb of listeners) cb();
  }
}

File: packages/py-core/ts_query/normalized_cache.py
"""Port of packages/core/src/normalized-cache.ts.

Pure data-structure work — no async, no framework — so this translates almost
line for line. Entities are stored flat ("User:42" -> fields) and queries hold
refs, so one ``write_fragment`` updates every query that touched the entity.
"""

from __future__ import annotations

from typing import Any, Callable, Dict, List, Optional, Set, Union

from .types import UNDEFINED, NormalizedCacheConfig, TypePolicy

EntityId = Union[str, int]


def _is_entity_ref(value: Any) -> bool:
    return isinstance(value, dict) and isinstance(value.get("__ref"), str)


def _is_plain_object(value: Any) -> bool:
    return isinstance(value, dict)


class NormalizedCache:
    # ##############################
    # INITIALIZATION
    # ##############################

    def __init__(self, config: Optional[NormalizedCacheConfig] = None) -> None:
        #: Flat entity store: "Type:id" -> merged fields
        self._entities: Dict[str, Dict[str, Any]] = {}
        #: Reverse index: entity ref -> query keys that reference it
        self._entity_to_queries: Dict[str, Set[str]] = {}
        #: Callbacks invoked when a specific entity changes (for use_fragment)
        self._entity_listeners: Dict[str, List[Callable[[], None]]] = {}
        self._config = config or NormalizedCacheConfig()

    # ##############################
    # PUBLIC API: SUBSCRIPTIONS
    # ##############################

    def subscribe_to_entity(
        self, typename: str, id: EntityId, callback: Callable[[], None]
    ) -> Callable[[], None]:
        """Subscribe to changes for one entity. Returns an unsubscribe callable."""
        ref = self._make_ref(typename, id)
        self._entity_listeners.setdefault(ref, []).append(callback)

        def unsubscribe() -> None:
            listeners = self._entity_listeners.get(ref)
            if listeners and callback in listeners:
                listeners.remove(callback)

        return unsubscribe

    # ##############################
    # PUBLIC API: NORMALIZATION
    # ##############################

    def normalize(self, data: Any, query_key: str) -> Any:
        """Store entities and return the normalized shape with entity refs."""
        return self._normalize_value(data, query_key)

    def denormalize(self, shape: Any) -> Any:
        """Reconstruct denormalized data from a normalized shape."""
        return self._denormalize_value(shape)

    # ##############################
    # PUBLIC API: FRAGMENT OPERATIONS
    # ##############################

    def write_fragment(self, typename: str, id: EntityId, data: Dict[str, Any]) -> List[str]:
        """Merge ``data`` into an entity; return query keys referencing it."""
        ref = self._make_ref(typename, id)
        existing = self._entities.get(ref, {})
        merged = self._merge_entity(typename, existing, data)
        self._entities[ref] = merged
        self._notify_entity_listeners(ref)
        return self._get_affected_queries(ref)

    def read_fragment(self, typename: str, id: EntityId) -> Any:
        """Read a raw entity record, or UNDEFINED when not cached."""
        return self._entities.get(self._make_ref(typename, id), UNDEFINED)

    def evict(self, typename: str, id: EntityId) -> List[str]:
        """Remove an entity; return the query keys that were referencing it."""
        ref = self._make_ref(typename, id)
        affected = self._get_affected_queries(ref)
        self._entities.pop(ref, None)
        self._entity_to_queries.pop(ref, None)
        self._notify_entity_listeners(ref)
        return affected

    # ##############################
    # INTERNAL HELPERS: ENTITY REFS
    # ##############################

    def _make_ref(self, typename: str, id: EntityId) -> str:
        return f"{typename}:{id}"

    def _get_entity_ref(self, obj: Dict[str, Any]) -> Optional[str]:
        """Derive an entity ref from __typename + key field(s), or None."""
        typename = obj.get("__typename")
        if not isinstance(typename, str):
            return None

        policy: Optional[TypePolicy] = self._config.type_policies.get(typename)
        key_fields: Union[str, Any] = (policy.key_fields if policy else None) or "id"

        if isinstance(key_fields, str):
            key_value = obj.get(key_fields)
        else:
            parts = [obj.get(f) for f in key_fields]
            if any(p is None for p in parts):
                return None
            key_value = ":".join(str(p) for p in parts)

        if key_value is None:
            return None
        return self._make_ref(typename, key_value)

    # ##############################
    # INTERNAL HELPERS: ENTITY MERGING
    # ##############################

    def _merge_entity(
        self, typename: str, existing: Dict[str, Any], incoming: Dict[str, Any]
    ) -> Dict[str, Any]:
        policy = self._config.type_policies.get(typename)
        if policy and policy.merge:
            return policy.merge(existing, incoming)
        return {**existing, **incoming}

    # ##############################
    # INTERNAL HELPERS: INDEX MANAGEMENT
    # ##############################

    def _get_affected_queries(self, ref: str) -> List[str]:
        return list(self._entity_to_queries.get(ref, ()))

    # ##############################
    # CORE LOGIC: NORMALIZATION
    # ##############################

    def _normalize_value(self, value: Any, query_key: str) -> Any:
        # Handle sequences
        if isinstance(value, (list, tuple)):
            return [self._normalize_value(item, query_key) for item in value]

        # Handle primitives and non-plain objects
        if not _is_plain_object(value):
            return value

        # Normalize nested fields
        normalized = {k: self._normalize_value(v, query_key) for k, v in value.items()}

        # Check if this object is an identifiable entity
        ref = self._get_entity_ref(value)
        if ref is None:
            # Not an identifiable entity — store inline
            return normalized

        # Store entity and build the reverse index
        existing = self._entities.get(ref, {})
        typename = value["__typename"]
        self._entities[ref] = self._merge_entity(typename, existing, normalized)
        self._entity_to_queries.setdefault(ref, set()).add(query_key)

        return {"__ref": ref}

    # ##############################
    # CORE LOGIC: DENORMALIZATION
    # ##############################

    def _denormalize_value(self, value: Any) -> Any:
        if isinstance(value, (list, tuple)):
            return [self._denormalize_value(item) for item in value]

        if not _is_plain_object(value):
            return value

        if _is_entity_ref(value):
            entity = self._entities.get(value["__ref"])
            if entity is None:
                return UNDEFINED  # entity was evicted
            return self._denormalize_value(entity)

        return {k: self._denormalize_value(v) for k, v in value.items()}

    # ##############################
    # INTERNAL HELPERS: NOTIFICATIONS
    # ##############################

    def _notify_entity_listeners(self, ref: str) -> None:
        for callback in list(self._entity_listeners.get(ref, ())):
            callback()
